expression/builtin_compare.go

```diff
@@ -981,17 +1014,13 @@ func compareReal(op opcode.Op, args []Expression, row []types.Datum, ctx context
 	if isNull0 || isNull1 {
 		if op == opcode.NullEQ {
 			if isNull0 && isNull1 {
-				return oneI64, false, nil
+				return 0, false, nil
 			}
-			return zeroI64, false, nil
+			return -1, false, nil
 		}
-		return zeroI64, true, nil
-	}
-	ret := resOfCmp(types.CompareFloat64(arg0, arg1), op)
-	if ret == -1 {
-		return zeroI64, false, errInvalidOperation.Gen("invalid op %v in comparison operation", op)
+		return -1, true, nil
 	}
-	return ret, false, nil
+	return int64(types.CompareFloat64(arg0, arg1)), false, nil
 }
 
 func compareDecimal(op opcode.Op, args []Expression, row []types.Datum, ctx context.Context) (val int64, isNull bool, err error) {
```

Comment:
You pass op only for check if it is NullEQ. You can consider this case only for NullEQ signature.